dags/oup/repository.py

```diff
@@ -0,0 +1,61 @@
+import os
+from collections import defaultdict
+from io import BytesIO
+from typing import IO
+
+from common.repository import IRepository
+from common.s3_service import S3Service
+
+
+class OUPRepository(IRepository):
+    ZIPED_DIR: str = "raw/"
+    EXTRACTED_DIR: str = "extracted/"
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.s3 = S3Service(os.getenv("OUP_BUCKET_NAME", "oup"))
+
+    def get_all_raw_filenames(self):
+        return [
+            os.path.basename(f.key)
+            for f in self.s3.objects.filter(Prefix=self.ZIPED_DIR).all()
+        ]
+
+    def find_all(self, filenames_to_process=None):
+        grouped_files = {}
+        filenames = (
+            filenames_to_process
+            if filenames_to_process
+            else self.__find_all_extracted_files()
+        )
+        if not filenames:
+            return []
+        for file in filenames:
+            last_part = os.path.basename(file)
+            filename_without_extension = last_part.split(".")[0]
+            grouped_files.setdefault(filename_without_extension, defaultdict(dict))
```

Comment:
one last thing, the comment was lost somewhere: do you really need a default dict there? It is a very good structure but takes at least twice of the memory as a standard `dict` does. In this use case I think you can just set a value of a dict as a simple dict `{value: key}`, or am I wrong?

Reply:
submitted the change :)